crates/ntx-builder/src/state/mod.rs

```diff
@@ -208,15 +232,24 @@ impl State {
 
         // Keep MMR pruned.
         let pruned_block_height =
-            (self.chain_mmr.chain_length().as_usize() - MAX_BLOCK_COUNT) as u32;
+            (self.chain_mmr.chain_length().as_usize().saturating_sub(MAX_BLOCK_COUNT)) as u32;
         self.chain_mmr.prune_to(..pruned_block_height.into());
     }
 
     /// Marks a previously selected candidate account as failed, allowing it to be available for
     /// selection again.
     #[instrument(target = COMPONENT, name = "ntx.state.candidate_failed", skip_all)]
-    pub fn candidate_failed(&mut self, candidate: NetworkAccountPrefix) {
-        self.in_progress.remove(&candidate);
+    pub fn candidate_failed(&mut self, candidate: &TransactionCandidate) {
+        let prefix = NetworkAccountPrefix::try_from(candidate.account.id()).unwrap();
```

Comment:
Could you add a `SAFETY` comment.

Reply:
Opted to pass the prefix into tx context instead.